File: 백준/Gold/2357. 최솟값과 최댓값/최솟값과 최댓값.java
import java.io.*;
import java.util.*;

public class Main {

    static final int MAX_N = 100000;
    static final int MAX_INT = Integer.MAX_VALUE;
    static final int MIN_INT = Integer.MIN_VALUE;

    static class Node {
        long minNode, maxNode;

        Node() {
            this.minNode = Long.MAX_VALUE; // 빈 노드는 최댓값으로 초기화
            this.maxNode = Long.MIN_VALUE; // 빈 노드는 최솟값으로 초기화
        }

        Node(long val) {
            this.minNode = val;
            this.maxNode = val;
        }

        Node(long minNode, long maxNode) {
            this.minNode = minNode;
            this.maxNode = maxNode;
        }
    }

    static long[] arr = new long[MAX_N];
    static Node[] tree = new Node[MAX_N * 4];

    static void init(int node, int start, int end) {
        if (start == end) {
            tree[node] = new Node(arr[start]);
            return;
        }

        int mid = (start + end) / 2;
        init(node * 2, start, mid);
        init(node * 2 + 1, mid + 1, end);

        long minVal = Math.min(tree[node * 2].minNode, tree[node * 2 + 1].minNode);
        long maxVal = Math.max(tree[node * 2].maxNode, tree[node * 2 + 1].maxNode);
        tree[node] = new Node(minVal, maxVal);
    }

    static Node findMinMax(int node, int start, int end, int left, int right) {
        if (right < start || end < left) {
            return new Node();
        }

        if (left <= start && end <= right) {
            return tree[node];
        }

        int mid = (start + end) / 2;
        Node LNode = findMinMax(node * 2, start, mid, left, right);
        Node RNode = findMinMax(node * 2 + 1, mid + 1, end, left, right);

        long minVal = Math.min(LNode.minNode, RNode.minNode);
        long maxVal = Math.max(LNode.maxNode, RNode.maxNode);

        return new Node(minVal, maxVal);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        // 각 줄마다 하나씩 수를 입력받음
        for (int i = 0; i < N; i++) {
            arr[i] = Long.parseLong(br.readLine());
        }

        // tree 배열 초기화
        for (int i = 0; i < tree.length; i++) {
            tree[i] = new Node();
        }

        init(1, 0, N - 1);

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            Node ans = findMinMax(1, 0, N - 1, a - 1, b - 1);
            sb.append(ans.minNode).append(" ").append(ans.maxNode).append("\n");
        }

        System.out.print(sb);
    }
}

File: 백준/Gold/2357. 최솟값과 최댓값/최솟값과 최댓값.cc
#include <iostream>
#include <algorithm>
using namespace std;

#define MAX_N 100000
#define MAX_INT INT32_MAX
#define MIN_INT INT32_MIN

struct Node
{
	long long minNode, maxNode;
	Node() : minNode(MAX_INT), maxNode(MIN_INT) {} // 기본 생성자
	Node(long long leafNode) : minNode(leafNode), maxNode(leafNode) {} // 단일 노드일 때 최소, 최대 초기화
	Node(long long _minNode, long long _maxNode) : minNode(_minNode), maxNode(_maxNode) {} // 최솟값과 최대값을 알고 있을 경우, 초기화
};

long long arr[MAX_N];
Node tree[MAX_N * 4];

void init(int node, int start, int end);
long long sectionSum(int node, int start, int end, int left, int right);
void update(int node, int start, int end, int index, long long val);

void init(int node, int start, int end)
{
	if (start == end) tree[node] = Node(arr[start]);
	else
	{
		init(node * 2, start, (start + end) / 2); // 왼쪽 노드
		init(node * 2 + 1, (start + end) / 2 + 1, end); // 오른쪽 노드
		tree[node] = Node(min(tree[node * 2].minNode, tree[node * 2 + 1].minNode), max(tree[node * 2].maxNode, tree[node * 2 + 1].maxNode));
	}
}

Node FindMinMax(int node, int start, int end, int left, int right)
{
	if (left > end || right < start) return Node(); 
	if (left <= start && end <= right) return tree[node]; 

	Node LNode = FindMinMax(node * 2, start, (start + end) / 2, left, right);
	Node RNode = FindMinMax(node * 2 + 1, (start + end) / 2 + 1, end, left, right);
	long long minNode = min(LNode.minNode, RNode.minNode);
	long long maxNode = max(LNode.maxNode, RNode.maxNode);

	return Node(minNode, maxNode);
}

int main()
{
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	int N, M, K;
	cin >> N >> M;

	for (int i = 0; i < N; i++) cin >> arr[i];

	init(1, 0, N - 1);// 트리 초기화 init(루트노드, 시작 인덱스, 끝 인덱스)

	for (int i = 0; i < M; i++)
	{
		long long a, b;
		cin >> a >> b;
		Node ans = FindMinMax(1, 0, N - 1, a - 1, b - 1);
		cout << ans.minNode << " " << ans.maxNode << "\n";										
	}
	return 0;
}